### DISCUSSION TITLE
엔티티의 생명주기에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by J-I-H-O** [2024-woowacourse-study/level-interview#153] [Spring Data]
![image](https://github.com/2024-woowacourse-study/level-interview/assets/110461155/b27177ef-d6d1-460d-bded-70adacfba020)

**Comment by jimi567** [comment]
엔티티의 생명 주기는 주로 다음과 같은 상태를 가지게 됩니다.

1. 비영속 상태 : 엔티티 객체가 생성되었지만, 아직 영속성 컨텍스트에 관리되지 않는 상태 DB와 아무런 관련이 없는 순수한 자바 객체
2. 영속 상태 : 영속성 컨텍스트에 관리되는 상태, flush가 호출되면 DB와 동기화 될 수 있고, 변경 사항이 영속성 컨텍스트의 변경 감지 등으로 반영되느 상태
3. 준영속 상태 : 영속성 컨텍스트에서 관리되다가 더 이상 관리 되지 않는 상태, 영속성 컨텍스트에서 detach 되거나, 영속성 컨텍스트가 clear 혹은 close 되면 준영속 상태가 된다.
4. 삭제 상태 : 영속성 컨텍스트에서 관리되다가 삭제된 상태

### 엔티티 상태 전이

* persist() : 비영속 상태에서 엔티티를 영속 상태로 전이
* merge() : 준영속 상태에서 영속 상태로 전이
* remove : 영속 상태를 삭제 상태로 전이
* detach() : 영속 상태에서 준영속 상태로 전이

---
번외 - 엔티티 생명주기 콜백 함수가 있다.
JPA는 엔티티 생명주기 동안 특정 이벤트가 발생할 때 호출될 수 있는 콜백 메서드를 제공합니다. 이러한 메서드는 @PrePersist, @PostPersist, @PreRemove, @PostRemove, @PreUpdate, @PostUpdate, @PostLoad 어노테이션을 사용하여 정의할 수 있습니다.

```java
@Entity
public class MyEntity {

    @Id
    private Long id;

    // 엔티티 필드들...

    @PrePersist
    public void prePersist() {
        System.out.println("Before persisting the entity");
    }

    @PostPersist
    public void postPersist() {
        System.out.println("After persisting the entity");
    }

   ...
}
```
위 처럼 어노테이션을 활용해, 특정 생명주기로 상태 변화 시 로직을 수행할 수 있게 됩니다.

**Comment by GIVEN53** [comment]
#### New 상태 (비영속)
```java
Member member = new Member("프린");
```
엔티티 객체가 새로 생성되었지만, 아직 영속성 컨텍스트와 연관되지 않은 상태입니다. 이 상태에서는 데이터베이스와 전혀 관련이 없습니다.
엔티티 객체는 메모리 상에만 존재합니다.

#### Managed 상태 (영속)
```java
em.persist(member);
em.merge(detagedMember);
em.find(Member.class, 1L);
```
엔티티 객체가 영속성 컨텍스트에 관리되고 있는 상태입니다. 이 상태에서는 엔티티의 변경 사항이 자동으로 데이터베이스에 반영됩니다.

#### Detached 상태 (준영속)
```java
em.detach(member);
em.clear();
em.close();
```
엔티티 객체가 한 번 영속성 컨텍스트에 의해 관리되었지만, 현재는 영속성 컨텍스트와 분리된 상태입니다. 이 상태에서는 엔티티 객체의 변경 사항이 더 이상 데이터베이스에 반영되지 않습니다.
영속성 컨텍스트 종료, 트랜잭션 종료 등으로도 준영속 상태로 전환됩니다.

#### Removed 상태 (삭제)
```java
em.remove(member);
```
엔티티 객체가 영속성 컨텍스트에서 제거된 상태입니다. 이 상태에서는 엔티티 객체가 데이터베이스에서 삭제됩니다.
